File: LeetCode/0341/cpp/lc0341.cpp
#include <cassert>
#include <stack>
#include <vector>

#include "nested_integer.hpp"

// Runtime complexity: O(n)
// Aux space complexity: O(n)
class NestedIterator {
public:
    NestedIterator(std::vector<NestedInteger>& nestedList) {
        addVectorNi(nestedList);
    };

    int next() {
        int ans = m_stack.top().getInteger();
        m_stack.pop();
        return ans;
    };

    bool hasNext() {
        // If stack is not empty and contains a list (so not isInteger()), process the list:
        while (!m_stack.empty() && !m_stack.top().isInteger()) {
            std::vector<NestedInteger> list = m_stack.top().getList();
            m_stack.pop();
            addVectorNi(list);
        }
        return !m_stack.empty();
    }

private:
    std::stack<NestedInteger> m_stack;
    void addVectorNi(const std::vector<NestedInteger>& vni) {
        for (int i = vni.size()-1; i >= 0; i--) {
            m_stack.push(vni[i]);
        }
    }
};

int main() {
}

File: LeetCode/0341/python/lc0341.py
class NestedIterator:
    def __init__(self, nestedList):
        # Convert nested list to a flat list.
        self.flat = []
        self.flatPos = 0;
        self.parse(nestedList)

    def parse(self, nestedList):
        for elem in nestedList:
            if (elem.isInteger()):
                self.flat.append(elem.getInteger())
            else:
                self.parse(elem.getList())

    def next(self):
        ans = self.flat[self.flatPos]
        self.flatPos += 1
        return ans

    def hasNext(self):
        return self.flatPos < len(self.flat)

